### ISSUE TITLE
Data race detected in metrics.Read

### ISSUE THREAD

**Issue by cristiangreco** [golang/go#53542]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18.3 darwin/arm64
</pre>

### Does this issue reproduce with the latest release?

Yes, still reproducible with 1.19beta1.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/Users/cristian/Library/Caches/go-build"
GOENV="/Users/cristian/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/cristian/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/cristian/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/opt/homebrew/Cellar/go/1.18.3/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/opt/homebrew/Cellar/go/1.18.3/libexec/pkg/tool/darwin_arm64"
GOVCS=""
GOVERSION="go1.18.3"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/98/fgzhqvmx7x1d61fqsvw9cg9m0000gn/T/go-build1308397819=/tmp/go-build -gno-record-gcc-switches -fno-common"
GOROOT/bin/go version: go version go1.18.3 darwin/arm64
GOROOT/bin/go tool compile -V: compile version go1.18.3
uname -v: Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:29 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T8101
ProductName:	macOS
ProductVersion:	12.4
BuildVersion:	21F79
lldb --version: lldb-1316.0.9.46
Apple Swift version 5.6.1 (swiftlang-5.6.0.323.66 clang-1316.0.20.12)
</pre></details>

### What did you do?

`runtime/metrics` says that "It is safe to execute multiple Read calls concurrently". However, running the following code with race detector enabled suggests a data race. Unsure whether this is a false positive (`metrics.Read` uses a semaphore lock).

```go
package main

import (
        "fmt"
        "runtime/metrics"
        "sync"
)

func main() {
        var wg sync.WaitGroup

        for i := 0; i < 2; i++ {
                wg.Add(1)
                go func() {
                        sample := make([]metrics.Sample, 1)
                        sample[0].Name = "/gc/heap/allocs:bytes"
                        metrics.Read(sample)
                        fmt.Printf("bytes allocated: %d\n", sample[0].Value.Uint64())
                        wg.Done()
                }()
        }

        wg.Wait()
}
```

### What did you expect to see?

No race detected.

### What did you see instead?

Data race detected.

```
$ go run -race main.go
bytes allocated: 511392
==================
WARNING: DATA RACE
Read at 0x00c000100030 by goroutine 8:
  runtime.evacuate_fast64()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/runtime/map_fast64.go:376 +0x3fc
  runtime/metrics.runtime_readMetrics()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/runtime/metrics.go:563 +0xc0

Previous write at 0x00c000100030 by goroutine 7:
  runtime.mapaccess2_faststr()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/runtime/map_faststr.go:108 +0x43c
  runtime.initMetrics()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/runtime/metrics.go:67 +0x254

Goroutine 8 (running) created at:
  main.main()
      /Users/cristian/Desktop/main.go:14 +0x50

Goroutine 7 (finished) created at:
  main.main()
      /Users/cristian/Desktop/main.go:14 +0x50
==================
bytes allocated: 552264
Found 1 data race(s)
exit status 66
```

**Comment by ianlancetaylor** [comment]
CC @golang/runtime

**Comment by prattmic** [comment]
I believe this is a false positive, as semacquire/semrelease aren't race instrumented.

cc @mknyszek

**Comment by aclements** [comment]
Given that readMetrics and initMetrics are both in package runtime (readMetrics is linknamed to runtime/metrics, but race instrumentation just looks at the package we're compiling), why did this get any race instrumentation at all?

**Comment by aclements** [comment]
Oh, readMetrics didn't get any race instrumentation. The race instrumentation in this case is hand-coded in the map implementation (and hence insensitive to what package it's being called from). Probably we just need a little manual race instrumentation around the map usage in the metrics code.

**Comment by aclements** [comment]
This isn't new in 1.19, so it doesn't have to be a release-blocker. It's also probably easy to fix, so for now I'll leave it marked just so we don't lose track.

**Comment by prattmic** [comment]
@gopherbot please backport to 1.17 and 1.18.

This triggers a false positive in the race detector for programs calling `runtime/metrics.Read` from multiple goroutines. There is a workaround: use a single global lock around all calls to `metrics.Read` in the program. However, that is quite invasive and has serious performance implications for a high-performance API.

**Comment by prattmic** [comment]
I don't think it is necessary. Maps aren't too widely used in the runtime, and I'd prefer us to be able to expand the number of races in the runtime rather than turning it off. It would be nice to catch these quicker though.

**Comment by cherrymui** [comment]
I recall Windows callback is another place that uses a map and causes false race, #50249 .

**Comment by prattmic** [comment]
Argh. I sent a fix for that too, but that one is nastier because it is called prior to scheduler initialization, which is fine for the raceread/racewrite in the map, but not raceacquire/racerelease.

**Comment by prattmic** [comment]
That one does me a bit more inclined to turn it off in the map.

**Comment by prattmic** [comment]
FWIW, I did a quick benchmark. A non-race map access is ~3ns. A race map access is ~17ns. With an additional runtime pc check (for `runtime.` prefix), a map access is ~32ns [1].

This isn't horrible given we're in race mode, but it is still pretty large.

[1] My map was in runtime_test, so the string comparison had to go through most of the characters before failing.

**Comment by cherrymui** [comment]
Thanks for the benchmark.

There are a few places where we check whether a PC is in the runtime. Maybe we could provide a 1-bit metadata for whether a function is a runtime function, to avoid the string comparison. But I guess in this case most of the overhead is probably the PC lookup, instead of the string comparison. Maybe we could make it simply an address range comparison, as the linker already lay out runtime functions together (mostly, I think).
